crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```
@@ -129,25 +149,222 @@ impl PgManager {
+        } else if let Some(before) = before {
+            let before = self.parse_obj_cursor(&before)?;
+            query = query
+                .filter(checkpoints::dsl::sequence_number.lt(before))
+                .order(checkpoints::dsl::sequence_number.desc());
+        }
+
+        let limit = first.or(last).unwrap_or(10) as i64;
+        query = query.limit(limit + 1);
+
+        let result: Option<Vec<StoredCheckpoint>> = self
+            .run_query_async(|conn| query.load(conn).optional())
+            .await?;
+
+        result
+            .map(|mut stored_checkpoints| {
+                let has_next_page = stored_checkpoints.len() as i64 > limit;
+                if has_next_page {
+                    stored_checkpoints.pop();
+                }
+
+                Ok((stored_checkpoints, has_next_page))
+            })
+            .transpose()
+    }
+
+    async fn multi_get_objs(
+        &self,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+        filter: Option<ObjectFilter>,
+    ) -> Result<Option<(Vec<StoredObject>, bool)>, Error> {
         let mut query = objects::dsl::objects.into_boxed();
-        query = query.filter(objects::dsl::object_id.eq(address));
 
-        if let Some(version) = version {
-            query = query.filter(objects::dsl::object_version.eq(version));
+        if let Some(filter) = filter {
+            if let Some(object_ids) = filter.object_ids {
+                query = query.filter(
+                    objects::dsl::object_id.eq_any(
+                        object_ids
+                            .into_iter()
+                            .map(|id| id.into_vec())
+                            .collect::<Vec<_>>(),
+                    ),
+                );
+            }
+
+            if let Some(_object_keys) = filter.object_keys {
+                // TODO: Temporary table? Probably better than a long list of ORs
```

If this isnt implemented yet, let it fail explicitly